// @codechecks/client/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// types for debug library suck so we dont use them
const debug = require("debug")("codechecks:client");
const chalk_1 = require("chalk");
const path_1 = require("path");
function printLogo() {
    console.log(`${chalk_1.default.magentaBright("Code")}${chalk_1.default.bold(chalk_1.default.magenta("Checks"))} ${chalk_1.default.yellow("Client")} ${chalk_1.default.dim(`v.${require("../package.json").version}`)}`);
    console.log(formatLink("https://codechecks.io"));
    console.log();
}
exports.printLogo = printLogo;
function bold(input) {
    return chalk_1.default.bold(input.toString());
}
exports.bold = bold;
function red(input) {
    return chalk_1.default.red(input.toString());
}
exports.red = red;
function green(input) {
    return chalk_1.default.green(input.toString());
}
exports.green = green;
function formatSHA(sha) {
    return sha.slice(0, 8);
}
exports.formatSHA = formatSHA;
function formatPath(path, rootPath) {
    return path_1.relative(rootPath, path);
}
exports.formatPath = formatPath;
function formatLink(link) {
    return chalk_1.default.underline(chalk_1.default.blueBright(link));
}
exports.formatLink = formatLink;
class Logger {
    log(...args) {
        console.log(...args);
    }
    debug(...args) {
        debug(...args);
    }
    error(...args) {
        console.error(chalk_1.default.red("Error occured: "), ...args);
    }
    critical(...args) {
        console.error(chalk_1.default.red("Critical error occured: "), ...args);
        process.exit(1);
    }
}
exports.logger = new Logger();
