// @tenderly/hardhat-tenderly/dist/tenderly/TenderlyService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenderlyService = exports.TENDERLY_DASHBOARD_BASE_URL = exports.TENDERLY_API_BASE_URL = void 0;
const index_1 = require("../index");
const TenderlyApiService_1 = require("./TenderlyApiService");
exports.TENDERLY_API_BASE_URL = "https://api.tenderly.co";
exports.TENDERLY_DASHBOARD_BASE_URL = "https://dashboard.tenderly.co";
class TenderlyService {
    static async getPublicNetworks() {
        let tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureAnonymousInstance();
        const apiPath = "/api/v1/public-networks";
        if (TenderlyApiService_1.TenderlyApiService.isAuthenticated()) {
            tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureInstance();
        }
        let response = [];
        try {
            response = (await tenderlyApi.get(apiPath)).data;
        }
        catch (e) {
            console.log(`Error in ${index_1.PluginName}: There was an error during the request. Network fetch failed`);
        }
        return response;
    }
    static async verifyContracts(request) {
        var _a;
        let tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureAnonymousInstance();
        const apiPath = "/api/v1/public/verify-contracts";
        if (TenderlyApiService_1.TenderlyApiService.isAuthenticated()) {
            tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureInstance();
        }
        try {
            const response = await tenderlyApi.post(apiPath, Object.assign({}, request));
            const responseData = response.data;
            let contract;
            if (responseData.bytecode_mismatch_errors != null) {
                console.log(`Error in ${index_1.PluginName}: Bytecode mismatch detected. Contract verification failed`);
                return;
            }
            if (!((_a = responseData.contracts) === null || _a === void 0 ? void 0 : _a.length)) {
                console.log(`${index_1.PluginName}: No new contracts have been verified`);
                return;
            }
            console.log("Smart Contracts successfully verified");
            console.group();
            for (contract of responseData.contracts) {
                const contractLink = `${exports.TENDERLY_DASHBOARD_BASE_URL}/contract/${index_1.ReverseNetworkMap[contract.network_id]}/${contract.address}`;
                console.log(`Contract ${contract.address} verified. You can view the contract at ${contractLink}`);
            }
            console.groupEnd();
        }
        catch (error) {
            console.log(`Error in ${index_1.PluginName}: There was an error during the request. Contract verification failed`);
        }
    }
    static async pushContracts(request, tenderlyProject, username) {
        var _a;
        const tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureInstance();
        try {
            const response = await tenderlyApi.post(`/api/v1/account/${username}/project/${tenderlyProject}/contracts`, Object.assign({}, request));
            const responseData = response.data;
            if (responseData.bytecode_mismatch_errors != null) {
                console.log(`Error in ${index_1.PluginName}: Bytecode mismatch detected. Contract push failed`);
                return;
            }
            if (!((_a = responseData.contracts) === null || _a === void 0 ? void 0 : _a.length)) {
                console.log(`${index_1.PluginName}: No new contracts have been pushed`);
                return;
            }
            const dashLink = `${exports.TENDERLY_DASHBOARD_BASE_URL}/${username}/${tenderlyProject}/contracts`;
            console.log(`Successfully pushed Smart Contracts for project ${tenderlyProject}. You can view your contracts at ${dashLink}`);
        }
        catch (error) {
            console.log(`Error in ${index_1.PluginName}: There was an error during the request. Contract push failed`);
        }
    }
}
exports.TenderlyService = TenderlyService;
